### PR TITLE
framework/media : Add BufferOutputDataSource.

### PR DESCRIPTION
**Pull request by YiHyunJin**
Data out through the buffer.
Callback should be used.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [26a180115000ab30d2e1f977460f8bd442c0ac79] - Code Rule Check Result:
framework/include/media/BufferOutputDataSource.h:44: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/include/media/BufferOutputDataSource.h:76: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:82: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:82: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)
framework/include/media/BufferOutputDataSource.h:82: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [26a180115000ab30d2e1f977460f8bd442c0ac79] - Code Rule Check (C++) OK.

File: framework/src/media/BufferOutputDataSource.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+bool BufferOutputDataSource::isPrepare()
+{
+	return true;
+}
+
+ssize_t BufferOutputDataSource::write(unsigned char* buf, size_t size)
+{
+	if (!buf) {
```
Comment (jsdosa) [inline framework/src/media/BufferOutputDataSource.cpp]: please, add condition for size

File: framework/src/media/BufferOutputDataSource.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		meddbg("buf is nullptr, hence return EOF\n");
+		return EOF;
+	}
+	ssize_t ret;
+	ret = mCallback(buf, size);
+	return ret;
```
Comment (jsdosa) [inline framework/src/media/BufferOutputDataSource.cpp]: How do we check if the user returns the wrong value ~?

File: framework/src/media/BufferOutputDataSource.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+ssize_t BufferOutputDataSource::write(unsigned char* buf, size_t size)
+{
+	if (!buf) {
+		meddbg("buf is nullptr, hence return EOF\n");
+		return EOF;
+	}
+	ssize_t ret;
+	ret = mCallback(buf, size);
```
Comment (jsdosa) [inline framework/src/media/BufferOutputDataSource.cpp]: Don't we need memcpy?

Comment (jsdosa) [reply framework/src/media/BufferOutputDataSource.cpp]: please check buffer life cycle
who delete the buffer? user or our framework?

Comment (YiHyunJin) [reply framework/src/media/BufferOutputDataSource.cpp]: @jsdosa  our frmework.

File: framework/src/media/BufferOutputDataSource.cpp (hunk 2 above)
Comment (YiHyunJin) [reply framework/src/media/BufferOutputDataSource.cpp]: @jsdosa I do not know what value the user returns. so... I do not know what to do.

File: framework/src/media/BufferOutputDataSource.cpp (hunk 1 above)
Comment (YiHyunJin) [reply framework/src/media/BufferOutputDataSource.cpp]: @jsdosa For example, are you talking about a negative value?

File: framework/src/media/BufferOutputDataSource.cpp (hunk 3 above)
Comment (Taejun-Kwon) [inline framework/src/media/BufferOutputDataSource.cpp]: mCallback & recorder_callback_t are not good for function name, kind of OnBufferDataReached will be enough.

File: framework/include/media/BufferOutputDataSource.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+ * @file media/BufferOutputDataSource.h
+ * @brief Media BufferOutputDataSource APIs
+ */
+
+#ifndef __MEDIA_BUFFEROUTPUTDATASOURCE_H
+#define __MEDIA_BUFFEROUTPUTDATASOURCE_H
+
+#include <media/OutputDataSource.h>
+
+typedef ssize_t(*recorder_callback_t)(void* buffer, size_t size);
```
Comment (Taejun-Kwon) [inline framework/include/media/BufferOutputDataSource.h]: Please add TODO here, based on my last proposal.

Comment (YiHyunJin) [reply framework/include/media/BufferOutputDataSource.h]: okay!

File: framework/src/media/BufferOutputDataSource.cpp (hunk 3 above)
Comment (YiHyunJin) [reply framework/src/media/BufferOutputDataSource.cpp]: Would you like to change recorder_callback_t to OnBufferDataReached?

**Comment by seinfra** [comment]
Target : [fdd3f68bacff1f156814247bed937422381e56a9] - Code Rule Check Result:
framework/include/media/BufferOutputDataSource.h:45: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/include/media/BufferOutputDataSource.h:77: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [fdd3f68bacff1f156814247bed937422381e56a9] - Code Rule Check (C++) OK.

File: framework/src/media/BufferOutputDataSource.cpp (hunk 2 above)
Comment (jsdosa) [reply framework/src/media/BufferOutputDataSource.cpp]: The mCallback function is a user callback.
What value should the user return?

File: framework/src/media/BufferOutputDataSource.cpp (hunk 1 above)
Comment (jsdosa) [reply framework/src/media/BufferOutputDataSource.cpp]: yes, you should check the value.

File: framework/src/media/BufferOutputDataSource.cpp (hunk 3 above)
Comment (Taejun-Kwon) [reply framework/src/media/BufferOutputDataSource.cpp]: yeah

**Comment by seinfra** [comment]
Target : [d6e8bd998943d343c4d18b771b1390ece42830f4] - Code Rule Check Result:
framework/include/media/BufferOutputDataSource.h:45: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/include/media/BufferOutputDataSource.h:77: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)
framework/include/media/BufferOutputDataSource.h:83: ERROR: [SPC_M_OPR] need consistent spacing around '&' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [d6e8bd998943d343c4d18b771b1390ece42830f4] - Code Rule Check (C++) OK.

File: framework/src/media/BufferOutputDataSource.cpp (hunk 1 above)
Comment (YiHyunJin) [reply framework/src/media/BufferOutputDataSource.cpp]: @jsdosa i fixed.

File: framework/include/media/BufferOutputDataSource.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	BufferOutputDataSource& operator=(const BufferOutputDataSource& source);
+	/**
+	 * @brief Deconstructs an empty BufferOutputDataSource.
+	 * @details @b #include <media/BufferOutputDataSource.h>
+	 * @since TizenRT v2.0 PRE
+	 */
+	virtual ~BufferOutputDataSource();
+
+	/**
+	 * @brief Whether file is ready to be write.
```
Comment (jsdosa) [inline framework/include/media/BufferOutputDataSource.h]: please, modify the comments.

**Pull request merged**